runtime/interpreter/value_range.go

```diff
@@ -0,0 +1,208 @@
+			LocationRange: locationRange,
+			Message:       fmt.Sprintf("sequence is moving away from end: %s due to the value of step: %s and start: %s", end, step, start),
+		})
+	}
+
+	fields := []CompositeField{
+		{
+			Name:  sema.InclusiveRangeTypeStartFieldName,
+			Value: start,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeEndFieldName,
+			Value: end,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeStepFieldName,
+			Value: step,
+		},
+	}
+
+	rangeSemaType := getInclusiveRangeSemaType(interpreter, rangeType)
+
+	rangeValue := NewCompositeValueWithStaticType(
+		interpreter,
+		locationRange,
+		nil,
+		rangeSemaType.QualifiedString(),
+		common.CompositeKindStructure,
+		fields,
+		common.ZeroAddress,
+		rangeType,
+	)
+
+	rangeValue.Functions = map[string]FunctionValue{
+		sema.InclusiveRangeTypeContainsFunctionName: NewHostFunctionValue(
+			interpreter,
+			sema.InclusiveRangeContainsFunctionType(
+				rangeSemaType.MemberType,
+			),
+			func(invocation Invocation) Value {
+				needleInteger := convertAndAssertIntegerValue(invocation.Arguments[0])
+
+				return rangeContains(
+					rangeValue,
+					rangeType,
+					invocation.Interpreter,
+					invocation.LocationRange,
+					needleInteger,
+				)
+			},
+		),
+	}
+
+	return rangeValue
+}
+
+func getInclusiveRangeSemaType(interpreter *Interpreter, rangeType InclusiveRangeStaticType) *sema.InclusiveRangeType {
+	return interpreter.MustConvertStaticToSemaType(rangeType).(*sema.InclusiveRangeType)
+}
+
+func rangeContains(
+	rangeValue *CompositeValue,
+	rangeType InclusiveRangeStaticType,
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	needleValue IntegerValue,
+) BoolValue {
+	start := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeStartFieldName)
+	endInclusive := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeEndFieldName)
+	step := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeStepFieldName)
+
+	var result bool
+	result = start.Equal(interpreter, locationRange, needleValue) ||
+		endInclusive.Equal(interpreter, locationRange, needleValue)
+
+	if !result {
+		greaterThanStart := needleValue.Greater(interpreter, start, locationRange)
+		greaterThanEndInclusive := needleValue.Greater(interpreter, endInclusive, locationRange)
+
+		if greaterThanStart == greaterThanEndInclusive {
+			// If needle is greater or smaller than both start & endInclusive, then it is outside the range.
+			result = false
+		} else {
+			// needle is in between start and endInclusive.
+			// start + k * step should be equal to needle i.e. (needle - start) mod step == 0.
+			diff, ok := needleValue.Minus(interpreter, start, locationRange).(IntegerValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			result = diff.Mod(interpreter, step, locationRange).Equal(interpreter, locationRange, interpreter.GetValueForIntegerType(0, rangeType.ElementType))
```

Comment:
Can break this long line into multiple lines:
```suggestion
			zeroValue := interpreter.GetValueForIntegerType(0, rangeType.ElementType)
			result = diff.Mod(interpreter, step, locationRange).Equal(interpreter, locationRange, zeroValue)
```